### ISSUE TITLE
Kill simulation after a crash

### ISSUE THREAD

**Issue by tomchor** [CliMA/Oceananigans.jl#1196]
I am trying to reproduce a simulation similar to the one in [this paper](https://journals.ametsoc.org/jpo/article/45/6/1554/12449/The-Energetics-of-Centrifugal-Instability) and when I checked the simulation after a whole day running I found it had crashed in the first hour. (Even though I was using the TimeStepWizard.) Or at least that's what I think happened judging by the output:

```
i:  13960, sim time: 11.103 hours, wall time: 42.767 minutes, Δt: 2.875 seconds, CFL: 1.00e+00
i:  13970, sim time: 11.111 hours, wall time: 42.797 minutes, Δt: 2.875 seconds, CFL: 1.00e+00
i:  13980, sim time: 11.119 hours, wall time: 42.827 minutes, Δt: 2.875 seconds, CFL: 1.00e+00
i:  13990, sim time: 11.127 hours, wall time: 42.856 minutes, Δt: 2.875 seconds, CFL: NaN
i:  14000, sim time:  NaN years, wall time: 42.885 minutes, Δt:  NaN years, CFL: NaN
i:  14010, sim time:  NaN years, wall time: 42.913 minutes, Δt:  NaN years, CFL: NaN
```

I wonder if it's possible to implement a check that kills the simulation/throws an error when it notices that it crashed. In my LES Fortran code I implemented something like that and it prevented me from wasting many core-hours simulating `NaN`s. Basically it was just something that checked if there were any `NaN`s in the domain and killed it otherwise.

Thoughts?

**Comment by ali-ramadhan** [comment]
Ah yes that's a great idea! Certainly would have helped me a few times as well.

We do currently have a NaN checker "diagnostic": https://github.com/CliMA/Oceananigans.jl/blob/master/src/Diagnostics/nan_checker.jl

But it would be better if it was added by default (and it's enough to just check a single grid point).

Was hoping to tag and release v0.44.2 today but I should be able to implement this quickly and add it into this next release.

**Comment by tomchor** [comment]
I wasn't aware of the `nan_checker`. Is the implementation described anywhere?
But yeah, if you prefer waiting an extra day and release the next version with that feature added by default I think that might actually be best too. I think that's the behavior most people (if not all) would want anyway.

**Comment by ali-ramadhan** [comment]
It really should be documented as part of https://clima.github.io/OceananigansDocumentation/stable/model_setup/diagnostics/ but right now it's a little awkward to use so I'll also clean it up a bit and add some docs for it.

But yeah once the PR is merged it seems like it'll be more of a behind-the-scenes feature that users hopefully won't have to worry about.

**Comment by francispoulin** [comment]
This all sounds great but just wanted to mention to @tomchor , I would be very interested in seeing your results and/or code when you have it up and running.  Centrifugal instability, or symmetric instability and some like to call it, is something that I had a PhD student work on (he just graduated), and I'm very interested to look at this with Oceananigans.

**Comment by navidcy** [comment]
This is a duplicate of https://github.com/CliMA/Oceananigans.jl/issues/938

**Comment by ali-ramadhan** [comment]
Ah you're right @navidcy. Probably should have left the issue open until we made NaN checking automatic but I'll make #1198 close both issues.

**Comment by navidcy** [comment]
#938 was open but it was forgotten and buried deep down in page 23 of the issues :)

**Comment by glwagner** [comment]
A better solution would resolve #1095 and add the appearance of NaNs as a "stop critierion", rather than using a diagnostic.

Things are confusing because we essentially have multiple interfaces to specifying callbacks.

**Comment by tomchor** [comment]
If you guys are going to follow that path, is there any way to implement the behavior now with the current state of the model, instead of having to wait for #1095 to be resolved?

**Comment by glwagner** [comment]
Yes. The simplest way is to define a function `progress(sim)` that throws an error if your simulation NaNs, and pass this to `Simulation` when constructing it:

https://github.com/CliMA/Oceananigans.jl/blob/35f749ef09fe021d13e8bad6433d889e8c39d6ca/src/Simulations/simulation.jl#L45-L46

**Comment by glwagner** [comment]
I often use `progress` to log the simulation iteration, time, etc. Because I calculate `maximum(abs, u)` for this, I also often add a line that throws an error if `max_u = maximum(abs, u)` `isnan`.

**Comment by glwagner** [comment]
You can also define a function `stop_if_nan(simulation)` that returns `true` if a NaN is detected, and then add this function to a simulation's stop criteria via `push!(simulation.stop_criteria, stop_if_nan)`.

A few ways to do the same thing, which is why resolving #1138 would be nice for users. 

Sorry if I implied that #1138 had to be resolved first. The problem right now is there are _too many_ ways to stop a simulation, which makes the user interface confusing.

**Comment by tomchor** [comment]
I see! I think given my level of Julia knowledge the first option (the progress function throwing an error one) may be easiest. I'll give it a shot. Thanks!

**Comment by tomchor** [comment]
I tried to use your `Diagnostics.NaNChecker`, but couldn't figure out how to do it. So implemented my own `nan` checker in the progress function, which works on CPUs, but I get following error when running on GPUs:

```
i:     10, sim time: 22.857 seconds, wall time: 52.770 seconds, Δt: 2.286 seconds, diff CFL: 1.14e-01, adv CFL: 6.87e-01
ERROR: LoadError: scalar getindex is disallowed
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] assertscalar(::String) at /glade/u/home/tomasc/.julia/packages/GPUArrays/ZxsKE/src/host/indexing.jl:41
 [3] getindex at /glade/u/home/tomasc/.julia/packages/GPUArrays/ZxsKE/src/host/indexing.jl:96 [inlined]
 [4] getindex at /glade/u/home/tomasc/.julia/packages/OffsetArrays/ExQCD/src/OffsetArrays.jl:271 [inlined]
 [5] getindex at ./subarray.jl:257 [inlined]
 [6] _getindex at ./abstractarray.jl:1100 [inlined]
 [7] getindex at ./abstractarray.jl:1060 [inlined]
 [8] _broadcast_getindex at ./broadcast.jl:614 [inlined]
 [9] _getindex at ./broadcast.jl:645 [inlined]
 [10] _broadcast_getindex at ./broadcast.jl:620 [inlined]
 [11] getindex at ./broadcast.jl:575 [inlined]
 [12] macro expansion at ./broadcast.jl:950 [inlined]
 [13] macro expansion at ./simdloop.jl:77 [inlined]
 [14] copyto! at ./broadcast.jl:949 [inlined]
 [15] copyto! at ./broadcast.jl:886 [inlined]
 [16] copy at ./broadcast.jl:862 [inlined]
 [17] materialize(::Base.Broadcast.Broadcasted{Base.Broadcast.DefaultArrayStyle{3},Nothing,typeof(isnan),Tuple{SubArray{Float64,3,OffsetArrays.OffsetArray{Float64,3,CUDA.CuArray{Float64,3}},Tuple{UnitRange{Int64},UnitRange{Int64},UnitRange{Int64}},false}}}) at ./broadcast.jl:837
```

This is my progress function:

```
advCFL = oc.Diagnostics.AdvectiveCFL(wizard)
difCFL = oc.Diagnostics.DiffusiveCFL(wizard)
start_time = time_ns()
function progress(sim)
    msg = @printf("i: % 6d, sim time: % 10s, wall time: % 10s, Δt: % 10s, diff CFL: %.2e, adv CFL: %.2e\n",
                  sim.model.clock.iteration,
                  prettytime(sim.model.clock.time),
                  prettytime(1e-9 * (time_ns() - start_time)),
                  prettytime(sim.Δt.Δt),
                  difCFL(sim.model),
                  advCFL(sim.model),
                  )

    if any(isnan.(oc.interior(model.velocities.u)))
        throw("Simulation crashed!")
    end
    return msg
end
```

Am I missing something? I'm assuming there is a much more intelligent way to do this.

**Comment by ali-ramadhan** [comment]
Ah I think you encountered an example of CUDA scalar operations: https://juliagpu.gitlab.io/CUDA.jl/usage/workflow/#UsageWorkflowScalar

This is when the GPU doesn't know how to do a certain operation quickly using GPU kernels, in this case `any(isnan.(oc.interior(model.velocities.u)))`, so it does it one-by-one on the CPU.

We disable CUDA scalar operations in Oceananigans.jl because it can lead to huge performance hits that can be hard to locate.

In your case I think it's because `interior(model.velocities.u)` returns a view into `u` without halos (but CUDA.jl doesn't always know how to broadcast over views into CuArrays).

I think it should work if you change it to

```julia
    if any(isnan, model.velocities.u.data.parent)
        throw("Simulation crashed!")
    end
```

Hopefully we can merge PR #1198 soon so you don't have to worry about this kind of stuff...

**Comment by tomchor** [comment]
Thanks, that seems to work! Although, like you said, I hope #1198 gets merged soon and makes it into v0.44.2! (which I'm assuming should be released soon)
